### PR TITLE
Address deprecations from persistence

### PR DESCRIPTION
**Pull request by greg0ire**
A backwards-compatibility layer has been added to persistence to help
consumers move to the new namespacing. It is based on class aliases,
which means the type declaration changes should not be a BC-break: types
are the same.
See doctrine/persistence#71

This means:
- using the new namespaces
- adding autoload calls for new types to types that may be extended and
use persistence types in type declarations of non-constructor methods,
so that signature compatibility is recognized by old versions of php.
More details on this at
https://dev.to/greg0ire/how-to-deprecate-a-type-in-php-48cf

### CODE REVIEW

File: lib/Doctrine/ODM/PHPCR/Mapping/Driver/XmlDriver.php [hunk 1]
```diff
@@ -19,10 +19,10 @@
 
 namespace Doctrine\ODM\PHPCR\Mapping\Driver;
 
-use Doctrine\Common\Persistence\Mapping\ClassMetadata;
-use Doctrine\Common\Persistence\Mapping\Driver\FileDriver;
 use Doctrine\Common\Persistence\Mapping\MappingException as DoctrineMappingException;
```
Comment (SenseException) [inline lib/Doctrine/ODM/PHPCR/Mapping/Driver/XmlDriver.php:22]: This one looks suspicious like a candidate for a namespace change.

Comment (greg0ire) [reply lib/Doctrine/ODM/PHPCR/Mapping/Driver/XmlDriver.php:22]: Oh right I somehow missed it

File: lib/Doctrine/ODM/PHPCR/Configuration.php [hunk 2]
```diff
@@ -383,3 +385,5 @@ public function getUuidGenerator()
             };
     }
 }
+
+interface_exists(MappingDriver::class);
```
Comment (dbu) [inline lib/Doctrine/ODM/PHPCR/Configuration.php:389]: is this line on purpose?

**Comment by dbu** [comment]
ah, sorry, read your blogpost now. then it makes sense. and you have the explanation in the commit message.

i assume this can still go wrong in rather confusing ways if a user manages to load the old interface before loading phpcr-odm classes that load the aliasing... 

can you please add something to the changelog for this? mention that we now support the separate persistence library, and to check the alias thing when you get the error from using the old interface and loading that before the alias gets defined (maybe with an example error message so people can recognize it)

**Comment by dbu** [comment]
actually, would it not be safest if the doctrine-common files that define those interfaces look if they see the persistence interface and then define themselves as alias instead of loading their own definition? instead of every using library having to be careful about it?

**Comment by dbu** [comment]
i fixed the styleci issues in 1.5 branch (and master) - can you rebase this branch please?

File: lib/Doctrine/ODM/PHPCR/Configuration.php (hunk 2 above)
Comment (greg0ire) [reply lib/Doctrine/ODM/PHPCR/Configuration.php:389]: Yes, see the linked blog post

**Comment by greg0ire** [comment]
> i assume this can still go wrong in rather confusing ways if a user manages to load the old interface before loading phpcr-odm classes that load the aliasing...

No, because the old file is precisely where the aliasing happens.

> actually, would it not be safest if the doctrine-common files that define those interfaces look if they see the persistence interface and then define themselves as alias instead of loading their own definition? instead of every using library having to be careful about it?

Autoloading everything by default means we can't have the deprecation messages, because we can no longer tell if the user tried to autoload the old type themselves.

File: lib/Doctrine/ODM/PHPCR/Configuration.php (hunk 2 above)
Comment (dbu) [reply lib/Doctrine/ODM/PHPCR/Configuration.php:389]: sorry, i read the link only afterward and commented on my comment, but forgot to close this.

**Comment by dbu** [comment]
ah, i see. yeah BC and shared libraries are really tough to get right...

thanks a lot for the effort!

File: lib/Doctrine/ODM/PHPCR/Proxy/ProxyFactory.php [hunk 3]
```diff
@@ -29,6 +28,7 @@
 use Doctrine\Common\Util\ClassUtils;
 use Doctrine\ODM\PHPCR\DocumentManagerInterface;
 use Doctrine\ODM\PHPCR\Mapping\ClassMetadata;
+use Doctrine\Persistence\Mapping\ClassMetadata as BaseClassMetadata;
```
Comment (dbu) [inline lib/Doctrine/ODM/PHPCR/Proxy/ProxyFactory.php:31]: i wonder if this really works with php < 7.4.

we use this class in the skipClass method and that method is defined by the base class with the common persistence class metadata. phpstorm tells me that restricting the class will only work from php 7.4 on

Comment (greg0ire) [reply lib/Doctrine/ODM/PHPCR/Proxy/ProxyFactory.php:31]: I don't understand what you mean by "restricting the class", can you please elaborate?

**Comment by greg0ire** [comment]
Next up: adding support for persistence 2 and common 3.

**Comment by dbu** [comment]
we seem to also use the deprecated common proxy thing.

master is the 2.0 branch - i suggest that we could focus on upgrades there, it would also allow for BC breaks if necessary, which could make the upgrading a lot easier and cleaner.

File: lib/Doctrine/ODM/PHPCR/Proxy/ProxyFactory.php (hunk 3 above)
Comment (dbu) [reply lib/Doctrine/ODM/PHPCR/Proxy/ProxyFactory.php:31]: this factory extends a base class that declares the skipClass method but with Doctrine\Common\Persistence\Mapping\ClassMetadata as argument.

i thought PHP 7.4 allows this, but according to https://3v4l.org/rCbc9 all PHP 7 emit a warning, while PHP 8 will fatal on this.

**Comment by greg0ire** [comment]
I made a PR to 1.5, since I can't see any BC-breaks in this : #816

**Pull request merged**